[RDY] msgpack: Recognize buffers, windows, tabpages, and notifications.

splinterofchaos opened the pull request:
One day it dawned on me that we could start an embedded nvim instance with `rpcstart()`

``` vim
let vim = rpcstart('nvim', ['--embed'])
```

and send commands from within the parent nvim instance!

``` vim
call rpcrequest(vim, 'vim_set_current_line', 'SOME TEXT')
echo rpcrequest(vim, 'vim_get_current_line')  " Prints SOME TEXT
```

but that attempting to get a list of windows, buffers, or tabpages would hit an `abort()` from `object_to_vim()`.

``` vim
echo rpcrequest(vim, 'vim_get_buffers')
echo rpcrequest(vim, 'vim_get_tabpages')
echo rpcrequest(vim, 'vim_get_windows')
```

This PR makes the above lines not trigger that abort. Inspired by [vim-misc's xolox#misc#os#exec()](https://github.com/xolox/vim-misc/blob/master/autoload/xolox/misc/os.vim#L82), which spawns another vim instance in order to execute commands asyncronously.

Commit:
object_to_vim: Recognize buffers, windows, tabpages.

Commit:
msgpack: Allow notifications to execute commands.

Consider: `let vim = rpcstart('nvim', ['--embed'])`

Allows `rpcnotify(vim, ...)` to work like an asynchronous
`rpcrequest(nvim, ...)`.

Helped-by: Michael Reed <m.reed@mykolab.com>
Helped-by: Justin M. Keyes <>

splinterofchaos commented:
Sorry for the long delay. Rebased cleanly and did a little more factoring. I hope to have this in a good state in the next few days.

Commit:
channel: recognized nvim-style errors

splinterofchaos commented:
Should have marked this WIP earlier, but oh well, now it's RFC.

I added some error checking code because I realized that if one tried to send a bad request to the embedded process, nvim would improperly interpret it. I don't think there's anything else to add to this.

tarruda commented:
LGTM.

This PR will let us start using Neovim for scripting remote instances without depending on python. Perhaps even scripts like [this](http://www.reddit.com/r/neovim/comments/30kvl3/ive_been_using_the_experimental_terminal_mode_and/cpwr5gv) will be possible after this PR.

bfredl commented:
There seems to be some limitations:
```
let vim = rpcstart('./build/bin/nvim', ['--embed'])
let x = rpcrequest(vim, 'vim_get_current_buffer')
let y = rpcrequest(vim, 'buffer_get_name', x)
```
the last call fails with `wrong type for arg 1, expecting Buffer`. It seems that we coerce api objects to int ids, but then don't allow int ids as object arguments?

tarruda commented:
> the last call fails with wrong type for arg 1, expecting Buffer. It seems that we coerce api objects to int ids, but then don't allow int ids as object arguments?

This is a really bad vimscript limitation, there's no way to create "userdata objects" that wrap arbitrary C data or pointers(It's the reason why job ids are used in the job API).

One ugly workaround is to use dicts to represent these other types: If a dict contains a reserved key(eg: $type) it will be coerced to the corresponding type. The ideal would be adding a VAR_USERDATA type.

bfredl commented:
> The ideal would be adding a VAR_USERDATA type.

Would this be hard to andd?  They should be designed such that they could be easily implemented as lua[jit] userdata for the vimL->lua translation later on. I would imagine the data would be the actual id/pointer, and the metatable reference would represent a subtype. For now (vanilla vimscript) they could be a pointer/id field + a subtype enum?

tarruda commented:
> Would this be hard to add? They should be designed such that they could be easily implemented as lua[jit] userdata for the vimL->lua translation later on. I would imagine the data would be the actual id/pointer, and the metatable reference would represent a subtype. For now (vanilla vimscript) they could be a pointer/id field + a subtype enum?

At first glance it would be easy, but I'm not sure how it affects existing operators or functions. The current vimscript implementation is just plain horrible(strings are just executed, there's not even an AST), so there could be problems in adding a new type without reviewing every code that evaluates expressions. This will be much easier and predictable in @ZyX-I viml->lua branch.

bfredl commented:
The "simple" (but less typesafe) alternative would be to allow an int wherever an api object is expected (or perhaps have a mode where this is allowed). 

Another alternative, that would be significtantly more work but perhaps worth it (probably also easier after the transation), would be to autogenerate vimscript wrappers for all nvim api functions. these could also be used to call api functions on the _current_ nvim instance (so that new core features only need to add a new-style api, and have a resonable vimscript api generated)

tarruda commented:
> Another alternative, that would be significtantly more work but perhaps worth it (probably also easier after the transation), would be to autogenerate vimscript wrappers for all nvim api functions. these could also be used to call api functions on the current nvim instance (so that new core features only need to add a new-style api, and have a resonable vimscript api generated)

This is exactly what @ZyX-I did in his branch: https://github.com/ZyX-I/neovim/blob/luaviml/scripts/msgpack-gen.lua#L405-L420 and https://github.com/ZyX-I/neovim/blob/luaviml/src/nvim/viml/executor/executor.c#L111. 

He did it because AFAIK the new vimscript implementation uses the API to control Neovim, but I'm not sure I understand how this solves the current problem which is coercion of native types. Could you give an example?

bfredl commented:
Oh, I was meaning that the wrappers could translate from VAR_INT to the correct msgpack api type, but of cource, once we are in lua land we could just use usertypes anyway, nevermind.

bfredl commented:
So what about just allowing coercing to/from int for now? With a big warning that only passing these to a nvim msgpack function is well defined and one must NOT do any integer operation on these (why would one, anyway?)  as this will be unsupported in the future.

splinterofchaos commented:
> There seems to be some limitations:

Yes, I noticed this last night on a script to test out the new functionality: https://gist.github.com/splinterofchaos/1c6c8882e28185a98017

One test: Copy all buffers from the remote instance. What I'd like to write is this:
```vim
function! nasync#copy_bufs(vim)
  let save_async = a:vim.async
  let a:vim.async = 0  " Execute synchronously
  for b in a:vim.vim('get_buffers')
    new
    let lines = a:vim.buf('line_count', b)
    call append(0, a:vim.buf('get_line_slice', b, 0, lines, 1, 1))
      
  endfor
  let a:vim.async = save_async
endfunction
```
This is what I ended up writing due to this restriction
```vim
function! nasync#copy_bufs(vim)
  let save_async = a:vim.async
  let a:vim.async = 0
  call a:vim.wait()
  for b in a:vim.vim('get_buffers')

    call a:vim.cmd('b ' . string(b))              " Switch to buffer b.
    call a:vim.wait()
    let nlines = a:vim.eval('line("$")')  " Get the line count
    call a:vim.vim('input', 'gg')         " Go to the top line

    let fname = a:vim.eval('bufname("%")')
    echo "Copying " . fname
    new

    " Append each line to the current buf.
    let lnum = 1
    while lnum <= nlines
      call a:vim.wait()
      call append(line("$"), a:vim.vim('get_current_line'))
      call a:vim.vim('input', 'j')
      let lnum += 1
    endwhile
  endfor
  let a:vim.async = save_async
endfunction
```
and it doesn't actually work, reporting that every buffer except the last doesn't exist. It does copy the last buffer, though.
 
> So what about just allowing coercing to/from int for now?

That's where I'm leaning right now. As it is, the first commit coerces buffers to `VAR_NUMBER`s, so seems logical to me to do the other way around, too.

tarruda commented:
> So what about just allowing coercing to/from int for now?

:+1: Here's a patch that should make integers acceptable for functions that receive buffers/windows/tabpages

```diff
diff --git a/scripts/msgpack-gen.lua b/scripts/msgpack-gen.lua
index 9ff9b4c..5ad2486 100644
--- a/scripts/msgpack-gen.lua
+++ b/scripts/msgpack-gen.lua
@@ -183,13 +183,20 @@ for i = 1, #functions do
     local converted, convert_arg, param, arg
     param = fn.parameters[j]
     converted = 'arg_'..j
-    if real_type(param[1]) ~= 'Object' then
-      output:write('\n  if (args.items['..(j - 1)..'].type != kObjectType'..real_type(param[1])..') {')
+    local rt = real_type(param[1])
+    if rt ~= 'Object' then
+      output:write('\n  if (args.items['..(j - 1)..'].type == kObjectType'..rt..') {')
+      output:write('\n    '..converted..' = args.items['..(j - 1)..'].data.'..rt:lower()..';')
+      if rt:match('^Buffer$') or rt:match('^Window$') or rt:match('^Tabpage$') then
+        -- accept positive integers for Buffers, Windows and Tabpages
+        output:write('\n  } else if (args.items['..(j - 1)..'].type == kObjectTypeInteger && args.items['..(j - 1)..'].data.integer > 0) {')
+        output:write('\n    '..converted..' = (unsigned)args.items['..(j - 1)..'].data.integer;')
+      end
+      output:write('\n  } else {')
       output:write('\n    snprintf(error->msg, sizeof(error->msg), "Wrong type for argument '..j..', expecting '..param[1]..'");')
       output:write('\n    error->set = true;')
       output:write('\n    goto cleanup;')
-      output:write('\n  }')
-      output:write('\n  '..converted..' = args.items['..(j - 1)..'].data.'..real_type(param[1]):lower()..';\n')
+      output:write('\n  }\n')
     else
       output:write('\n  '..converted..' = args.items['..(j - 1)..'];\n')
     end

```

tarruda committed:
msgpack: coerce Ints to Windows/Buffers/Tabpages

Add conversion checking when generating msgpack handlers.

splinterofchaos commented:
@tarruda thanks! That's just it. I amended it to also accept `Booleans`. It's really fast, too!

Now that I've got both formal and informal tests going, I feel pretty confident about this PR. It bothers me that notifications can't receive error messages (which is why my script has an `async` variable), but I don't know this is a very solvable problem since the specification gives no way for clients to identify which notification an error came from.

Anyone interested in testing this PR, please check out my test script (linked below). It has some functions at the bottom for testing. It spawns `v:progpath` instead of just `nvim` so this version needn't be installed in order to work.

https://gist.github.com/splinterofchaos/1c6c8882e28185a98017

bfredl commented on test/functional/api/server_requests_spec.lua:
```diff
@@ -115,4 +115,34 @@ describe('server -> client', function()
+    it('can communicate buffers, tabpages, and windows', function()
+      eq({2}, eval("rpcrequest(vim, 'vim_get_buffers')"))
+      eq({3}, eval("rpcrequest(vim, 'vim_get_tabpages')"))
+      eq({1}, eval("rpcrequest(vim, 'vim_get_windows')"))
```
Perhaps also test  to pass an api object, like `buffer_set_line_slice` or something?

splinterofchaos replied:
:+1:

bfredl commented:
LGTM great work! though I have to use explicitly `./build/bin/nvim`, `v:progpath` is just `nvim` for me and gives installed version (but `/proc/self/exe` work nice :) )

Commit:
docs: update for rpcstart('nvim', ['--embed'])

Also, edit the reference to the non-existent api function,
vim_list_buffers, to vim_get_buffers.

splinterofchaos commented:
I'll leave this open for a day or so for final comments, but otherwise I'm ready to merge.

I added a little to the docs about this feature, though I could maybe say a little more.

Pyrohh commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -177,6 +183,10 @@ Buffer				  -> enum value kObjectTypeBuffer
 Window				  -> enum value kObjectTypeWindow
 Tabpage				  -> enum value kObjectTypeTabpage
 
+An api method expecting one of these types may be passed an integer instead,
```
`api` -> `API`

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -116,6 +116,12 @@ functions can be called interactively:
 <
+One can also spawn and connect to an embedded nvim instance via |rpcstart()|
+>
+    let vim = rpcstart('build/bin/nvim', ['--embed'])
```
We should probably not assume a development environment for examples, so maybe `/usr/bin/nvim` would be better.

splinterofchaos replied:
Oops -- :+1:

Pyrohh commented on src/nvim/api/private/defs.h:
```diff
@@ -22,6 +22,14 @@ typedef enum {
+  kMessageTypeNotification
+} MessageType;
+
+#define NO_RESPONSE UINT64_MAX
```
Perhaps this could be made into an enum? I've been doing this recently, and it's much nicer to debug IMO.

splinterofchaos replied:
A `const uint64_t` fine? Has to be the same type as message ID's.

Pyrohh replied:
Sure, really anything but macro :)

On April 11, 2015 6:33:46 PM EDT, Scott Prager <notifications@github.com> wrote:
>> @@ -22,6 +22,14 @@ typedef enum {
>>    kErrorTypeValidation
>>  } ErrorType;
>>  
>> +typedef enum {
>> +  kMessageTypeRequest,
>> +  kMessageTypeResponse,
>> +  kMessageTypeNotification
>> +} MessageType;
>> +
>> +#define NO_RESPONSE UINT64_MAX
>
>A `const uint64_t` fine? Has to be the same type as message ID's.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/1446/files#r28199006

Pyrohh commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -442,18 +460,18 @@ static void handle_request(Channel *channel, msgpack_object *request)
   // Retrieve the request handler
   MsgpackRpcRequestHandler handler;
-  msgpack_object method = request->via.array.ptr[2];
+  msgpack_object* method = msgpack_rpc_method(request);
```
pointers should be aligned to the right

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/msgpack_rpc/helpers.c:
```diff
@@ -351,49 +351,86 @@ void msgpack_rpc_serialize_response(uint64_t response_id,
   }
 }
 
+bool msgpack_rpc_is_notification(msgpack_object *req)
```
It looks like this can be made static.

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/msgpack_rpc/helpers.c:
```diff
@@ -351,49 +351,86 @@ void msgpack_rpc_serialize_response(uint64_t response_id,
+  return obj->type == MSGPACK_OBJECT_ARRAY ? obj : NULL;
+}
+
+msgpack_object *msgpack_rpc_msg_id(msgpack_object *req)
```
Same here regarding static

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/msgpack_rpc/helpers.c:
```diff
@@ -351,49 +351,86 @@ void msgpack_rpc_serialize_response(uint64_t response_id,
+  if (type == kMessageTypeRequest) {
+    msgpack_object *id_obj = msgpack_rpc_msg_id(req);
+    if (!id_obj) {
+      api_set_error(err, Validation, _("Id must be a positive integer"));
```
`Id` -> `ID`

splinterofchaos replied:
:+1:

Pyrohh commented on src/nvim/msgpack_rpc/helpers.c:
```diff
@@ -351,49 +351,86 @@ void msgpack_rpc_serialize_response(uint64_t response_id,
-  if (req->via.array.ptr[1].type != MSGPACK_OBJECT_POSITIVE_INTEGER) {
-    api_set_error(err, Validation, _("Id must be a positive integer"));
+  if (req->via.array.ptr[0].type != MSGPACK_OBJECT_POSITIVE_INTEGER) {
+    api_set_error(err, Validation, _("Message type must be an integer"));
```
Why was this changed? I'm not totally familiar with the code, but it seems odd given that the comparison above is with `MSGPACK_OBJECT_POSITIVE_INTEGER`...

splinterofchaos replied:
The diff is confusing. Checking the ID got moved further down; this line checks that the message type is an integer.

By the msgpack specification, requests are in the form `[kMessageTypeRequest, ID, method, args]` and notifications: `[kMessageTypeNotification, method, args]`. Only requests have ID's so we need to validate the `kMessageType*` and array size before retrieving the ID.

Pyrohh replied:
Ah, I see. Thanks for explaining.

Pyrohh commented on src/nvim/msgpack_rpc/helpers.c:
```diff
@@ -351,49 +351,86 @@ void msgpack_rpc_serialize_response(uint64_t response_id,
+  if (msgpack_rpc_is_notification(req)) {
+    return NULL;
+  }
+  msgpack_object  *obj = &req->via.array.ptr[1];
```
extra space after `msgpack_object`

splinterofchaos replied:
:+1:

splinterofchaos commented on src/nvim/api/private/defs.h:
```diff
@@ -22,6 +22,16 @@ typedef enum {
+
+
+/// Used as the message ID of notifications.
+#define NO_RESPONSE UINT64_MAX
```
Sorry, @Pyrohh, lua's cimport can't handle `uint64_t`, `size_t`, or the `enum` version, and I don't know how to `#if NOT_CIMPORT` it.

Pyrohh replied:
Ah, oh well.

splinterofchaos commented:
@Pyrohh Did you have any other concerns?

Pyrohh commented:
No, that was everything. Looks pretty good.

On April 12, 2015 10:09:54 PM EDT, Scott Prager <notifications@github.com> wrote:
>@Pyrohh Did you have any other concerns?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/1446#issuecomment-92171851

tarruda commented:
Great work. Later we can add a vimscript library to wrap `rpc` functions for calling methods discovered in a child nvim instance, something like this:

```vim
let vim = remotevim#New(argv)
echo vim.eval(expr)
call vim.command(cmd)
```

splinterofchaos commented:
I was planning to do that as an external plugin based on my testing script, but if you think we should ship with it, I can send a PR once it's ready.

tarruda commented:
> I was planning to do that as an external plugin based on my testing script, but if you think we should ship with it, I can send a PR once it's ready.

Not sure if this should be included, cc @justinmk
